File: day2/javascript/solution-part-two.js
const { readFile, every } = require("./utils.js");

const lines = readFile("./input.txt");
let sumOfPowers = 0;

for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    let colors = {
        red: 0,
        blue: 0,
        green: 0,
    };

    let [id, rounds] = line.split(":");
    id = parseInt(id.split(" ")[1]);

    let maxSet = {
        red: 0,
        blue: 0,
        green: 0,
    };

    for (const round of rounds.split(";")) {
        const data = every(round.split(","), (x) => x.trim());
        for (const type of data) {
            const [val, color] = type.split(" ");
            const num = parseInt(val);
            if(num > colors[color]) colors[color] = num
        }

        maxSet.red = Math.max(maxSet.red, colors.red);
        maxSet.green = Math.max(maxSet.green, colors.green);
        maxSet.blue = Math.max(maxSet.blue, colors.blue);
    }

    const power = maxSet.red * maxSet.green * maxSet.blue;
    sumOfPowers += power;
}

console.log("Sum of Powers:", sumOfPowers);

File: day2/python/solution-part-two.py
def main():
    with open("input.txt", "r") as file:
        lines = file.readlines()

    sum_of_powers = 0

    for line in lines:
        colors = {"red": 0, "blue": 0, "green": 0}
        max_set = {"red": 0, "blue": 0, "green": 0}

        parts = line.split(":")
        game_id = int(parts[0].split(" ")[1])

        for round_data in parts[1].split(";"):
            data = [x.strip() for x in round_data.split(",")]

            for item in data:
                val, color = item.split(" ")
                num = int(val)
                if num > colors[color]:
                    colors[color] = num

            max_set["red"] = max(max_set["red"], colors["red"])
            max_set["green"] = max(max_set["green"], colors["green"])
            max_set["blue"] = max(max_set["blue"], colors["blue"])

        power = max_set["red"] * max_set["green"] * max_set["blue"]
        sum_of_powers += power

    print("Sum of Powers:", sum_of_powers)

if __name__ == "__main__":
    main()
